src/reify.ml4

```diff
@@ -724,6 +911,58 @@ struct
+    let (evm,pgm) = reduce_hnf env (evm, pgm) in 
+    let (coConstr,args) = app_full pgm [] in
+    if Term.eq_constr coConstr tmReturn then
+      match args with
+      | _::h::[] -> (env,evm,h)
+      | _ -> raise (Failure "tmReturn must take 2 arguments. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmBind then
+      match args with
+      | _::_::a::f::[] ->
+        let (env, evm, ar) = run_template_program_rec (env,evm,a) in
+        run_template_program_rec (env,evm,(Term.mkApp (f, Array.of_list [ar])))
+      | _ -> raise (Failure "tmBind must take 4 arguments. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmMkDefinition then
+      match args with
+      | b::name::_::body::[] -> 
+        let (evm,name) = reduce_all env (evm,name) in
+        let (evm,b) = reduce_all env (evm,b) in
+        let _ = unquote_red_add_definition (from_bool b) env evm (unquote_ident name) body in (env, evm, unit_tt)
+      | _ -> raise (Failure "tmMkDefinition must take 4 arguments. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmQuote then
+      match args with
+      | id::b::[] ->
+          let (evm,id) = reduce_all env (evm,id) in
+          let (evm,b) = reduce_all env (evm,b) in
+          let qt=quote_decl (from_bool b) env evm (unquote_string id) in
+          (env, evm, qt)
+      | _ -> raise (Failure "tmQuote must take 1 argument. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmQuoteTerm then
+      match args with
+      | _::trm::[] -> let qt = quote_term env trm in (* user should do the reduction (using tmReduce) if they want *)
+              (env, evm, qt)
+      | _ -> raise (Failure "tmQuoteTerm must take 1 argument. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmQuoteTermRec then
+      match args with
+      | trm::[] -> let qt = quote_term_rec env trm in
+              (env, evm, qt)
+      | _ -> raise (Failure "tmQuoteTermRec must take 1 argument. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmPrint then
+      match args with
+      | _::trm::[] -> let _ = Pp.msg_debug ((Printer.pr_constr trm)) in (env, evm, unit_tt)
+      | _ -> raise (Failure "tmPrint must take 2 arguments. Please file a bug with Template-Coq.")
+    else if Term.eq_constr coConstr tmReduce then
+      match args with
+      | _(*reduction strategy*)::_(*type*)::trm::[] -> 
+          let (evm,trm) = reduce_all env (evm,trm) in (env, evm, trm)
+      | _ -> raise (Failure "tmReduce must take 3 arguments. Please file a bug with Template-Coq.")
+    else raise (Failure "Invalid argument or yot yet implemented. The argument must be a TemplateProgram")
+
+  let run_template_program (env: Environ.env) (evm: Evd.evar_map) (body: Constrexpr.constr_expr) : unit =
+  	let (body,_) = Constrintern.interp_constr env evm body in
+    let _ = run_template_program_rec (env,evm,body) in ()
```

Comment:
Maybe we should check the type here? That might produce a better error message than delaying it until we get into the interpreter.